File: go/0299_bulls_and_cows.go
import "fmt"

func getHint(secret string, guess string) string {
	b, c := 0, 0

	ms, mg := make(map[byte]int), make(map[byte]int)
	for i := 0; i < len(secret); i++ {
		if cnt, ok := ms[secret[i]]; ok {
			ms[secret[i]] = cnt + 1
		} else {
			ms[secret[i]] = 1
		}

		if cnt, ok := mg[guess[i]]; ok {
			mg[guess[i]] = cnt + 1
		} else {
			mg[guess[i]] = 1
		}
	}

	for k, v := range ms {
		if cnt, ok := mg[k]; ok {
			c += min(cnt, v)
		}
	}

	for i := 0; i < len(secret); i++ {
		if secret[i] == guess[i] {
			b += 1
			c -= 1
		}
	}
	return fmt.Sprintf("%dA%dB", b, c)
}

func min(a, b int) int {
	if a >= b {
		return b
	}
	return a
}

File: python/0299_bulls_and_cows.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        smap, gmap = {}, {}
        for i, ch in enumerate(secret):
            if ch in smap:
                smap[ch][i] = True
            else:
                smap[ch] = {i: True}

        a = b = 0
        exactFound = set()
        for i, ch in enumerate(guess):
            if ch in smap and i in smap[ch]:
                a += 1
                del smap[ch][i]
                exactFound.add(i)

        for i, ch in enumerate(guess):
            if i not in exactFound and ch in smap and len(smap[ch]) > 0:
                b += 1
                k = None
                for key in smap[ch].keys():
                    k = key
                    break
                del smap[ch][k]

        return str(a) + 'A' + str(b) + 'B'
